vertical-pod-autoscaler/e2e/v1/recommender.go

```diff
@@ -226,6 +226,73 @@ var _ = RecommenderE2eDescribe("VPA CRD object", func() {
 	})
 })
 
+var _ = RecommenderE2eDescribe("VPA ContainerScalingMode", func() {
+	f := framework.NewDefaultFramework("vertical-pod-autoscaling")
+
+	var (
+		vpaClientSet *vpa_clientset.Clientset
+	)
+
+	ginkgo.BeforeEach(func() {
+		config, err := framework.LoadConfig()
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		vpaClientSet = vpa_clientset.NewForConfigOrDie(config)
+	})
+
+	ginkgo.It("ContainerScalingModeAuto recommendations", func() {
+		ginkgo.By("Setting up a hamster deployment")
+		createTwoHamestersDeployment(f)
+		vpaCRD := createVpaCRDWithContainerScalingMode(f, vpa_types.ContainerScalingModeAuto)
+
+		ginkgo.By("Waiting for recommendation to be filled")
+		vpa, err := WaitForRecommendationPresent(vpaClientSet, vpaCRD)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		gomega.Expect(len(vpa.Status.Recommendation.ContainerRecommendations)).Should(gomega.BeNumerically("==", 2))
+
+	})
+
+	ginkgo.It("ContainerScalingModeOff recommendations", func() {
+		ginkgo.By("Setting up a hamster deployment")
+		createTwoHamestersDeployment(f)
+		vpaCRD := createVpaCRDWithContainerScalingMode(f, vpa_types.ContainerScalingModeOff)
+
+		ginkgo.By("Waiting for recommendation to be filled")
+		vpa, err := WaitForRecommendationPresent(vpaClientSet, vpaCRD)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		// hamster container has recommendations turned off,
+		// so we expect only one container recommendation for hamster2.
+		gomega.Expect(len(vpa.Status.Recommendation.ContainerRecommendations)).Should(gomega.BeNumerically("==", 1))
+		gomega.Expect(vpa.Status.Recommendation.ContainerRecommendations[0].ContainerName).To(gomega.Equal("hamster2"))
+	})
+})
+
+func createTwoHamestersDeployment(f *framework.Framework) {
+	d := NewHamsterDeploymentWithResourcesAndLimits(f,
+		ParseQuantityOrDie("300m") /*cpu request*/, ParseQuantityOrDie("300Mi"), /*memory request*/
+		ParseQuantityOrDie("300m") /*cpu limit*/, ParseQuantityOrDie("300Mi") /*memory limit*/)
+	d.Spec.Template.Spec.Containers = append(d.Spec.Template.Spec.Containers, d.Spec.Template.Spec.Containers[0])
```

Comment:
You can use SetupHamsterContainer function as well https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/e2e/v1/common.go#L240

Reply:
I kept it as it is. Even with SetupHamsterContainer is till requires appending and setting name.

Reply:
Could we have something more general than having "Two" in the method name?

Reply:
That's good idea.
Although there is already "Two" convention in e2e common: https://github.com/kubernetes/autoscaler/blob/786a8ec8147586e98420c4433bda5141e76b55fc/vertical-pod-autoscaler/e2e/v1/common.go#L278
I think it requires refactor. I will add separate PR with some refactor and also createTwoHamstersDeployment added to e2e common.

Reply:
I've moved all code refactor to #2643.

Reply:
Done. #2643 merged.